presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,482 @@
+    }
+
+    public Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(
+                new PushDownDereferenceThroughJoin(metadata, sqlParser),
+                new PushDownDereferenceThroughSort(metadata, sqlParser),
+                new PushDownDereferenceThroughUnnest(metadata, sqlParser),
+                new PushDownDereferenceThroughProject(metadata, sqlParser));
+    }
+
+    private abstract class DereferencePushDownRule<N extends PlanNode>
+            implements Rule<ProjectNode>
+    {
+        private final Capture<N> targetCapture = newCapture();
+        private final Pattern<N> targetPattern;
+
+        protected final Metadata metadata;
+        protected final SqlParser sqlParser;
+
+        protected DereferencePushDownRule(Metadata metadata, SqlParser sqlParser, Pattern<N> targetPattern)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+            this.targetPattern = requireNonNull(targetPattern, "targetPattern is null");
+        }
+
+        @Override
+        public Pattern<ProjectNode> getPattern()
+        {
+            return project().with(source().matching(targetPattern.capturedAs(targetCapture)));
+        }
+
+        @Override
+        public Result apply(ProjectNode node, Captures captures, Context context)
+        {
+            N child = captures.get(targetCapture);
+            Map<DereferenceExpression, Symbol> expressions = getDereferenceSymbolMap(node.getAssignments().getExpressions().stream().collect(toImmutableList()), context, metadata, sqlParser);
+            Assignments assignments = node.getAssignments().rewrite(new DereferenceReplacer(expressions));
+
+            Result result = pushDownDereferences(context, child, expressions, assignments);
+            if (result.isEmpty()) {
+                return Result.empty();
+            }
+            return Result.ofPlanNode(new ProjectNode(context.getIdAllocator().getNextId(), result.getTransformedPlan().get(), assignments));
+        }
+
+        protected abstract Result pushDownDereferences(Context context, N targetNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments);
+
+        protected boolean isCastRowType(DereferenceExpression dereference)
+        {
+            Expression base = dereference.getBase();
+            if (base instanceof Cast) {
+                return metadata.getTypeManager().getType(parseTypeSignature(((Cast) base).getType())) instanceof RowType;
+            }
+            return false;
+        }
+    }
+
+    /** Transforms:
+     * <pre>
+     *  Project(a_x := a.msg.x)
+     *    Join(a_y = b_y) => [a]
+     *      Project(a_y := a.msg.y)
+     *          Source(a)
+     *      Project(b_y := b.msg.y)
+     *          Source(b)
+     *  </pre>
+     * to:
+     * <pre>
+     *  Join(a_y = b_y) => [a_x]
```

Comment:
Other examples show an identity projection in the result, but it is missing here. At the same time the code seems to be actually generating an identity projection and there is no logic to remove it. Let's add it here or add logic to remove it to the rule.